### ISSUE TITLE
Cart and Checkout Blocks: Add in option for toggling coupon code on or off.

### ISSUE THREAD

**Issue by nerrad** [woocommerce/woocommerce-blocks#2316]
This was scoped for MVP but got missed (caught by @garymurray):

![Image 2020-04-27 at 9.10.43 AM.png](https://images.zenhubusercontent.com/55c8bfc24f7a08fbb7f431bd/7664680a-ac89-491c-8d90-d0d70da2dba4)

Slack convo: (p1587985860288400-slack-C8X6Q7XQU):

Some remaining decision points:

- Do we have the setting existing on both blocks?
- Is the setting independent for both blocks (i.e. toggling coupons off for cart is independent for what happens in the checkout block). If not independent then it's more difficult to implement because the block attributes can't be the source for the toggle value. However, we could just have the toggle affect the global setting (possibly using REST api endpoint if it exists for this setting).

cc @garymurray for feedback.

**Comment by garymurray** [comment]
Thinking through why a merchant would want to be able to control this at the block level. The main reason merchants might want to remove them on one of the pages is to increase the conversion rate of their stores - with that in mind:

I'd imagine that earlier in the "normal" store setup process one chooses to define if you will/will not use coupons and enable/disable them accordingly. This is currently controlled via a core WooCommerce setting - so perhaps we leave it like that for now with regards to the **global** control of using them vs not using them.

When it comes to allowing a merchant to control it **at a block level** I think we should look to only show the option to control the coupons in the blocks if a merchant has chosen to allow coupons in their store via the core setting.

If they have, then they are given the option to control the use of coupons in either the cart and checkout pages on an 'independent' basis (so that the settings work independently of each other).

This still allows the merchant to accept coupons, but they decide if they would prefer them to be used on both the cart and checkout pages, or only one of them.

**Comment by nerrad** [comment]
I'm concerned that this ux might be too complicated for merchants and lead to paralysis on where they configure their coupon options for display in the cart or checkout blocks. Considering potential movement of coupon management to a dedicated admin page, I think we should for now _not_ make this a block level attribute but instead just have the toggle affect the global option, that way it's always in sync with either being set in the dedicated coupon management page (whether that's the existing global setting - or a future dedicated coupon admin page) or the merchant is setting up their cart and checkout blocks and decide to toggle on/off there.

With that said, I think we need a way in the block inspector tools to differentiate between _block-level settings_ and _global settings_ that are configurable via the blocks _for convenience_.

**Comment by garymurray** [comment]
So to understand - based on your comment:

- The block setting would inherit the WooCommerce core setting
- You can change that at the block level if you wanted, which changes it in the core setting
- If you change it in the cart block, it changes it in the core setting and in turn changes it in the checkout block so that if you then went to view the checkout block it would "inherit" the change you made in the cart block

As an example:

- Merchant sets the core setting to enable coupons
- Merchant then views the cart block and see coupons are enabled and decides to disable coupons (which then changes the core setting to 'disabled')
- Merchant then views the checkout block and sees that coupons are now disabled as per their previous choice

Is this what you are describing?

**Comment by nerrad** [comment]
> Is this what you are describing?

Correct. But that's also why I think we need some way to clearly differentiate in the ui settings that are "global" settings being changed vs block specific settings. A very _quick_ mockup to illustrate (but we'd probably have coupons in their own panel..simply to illustrate):

<img width="306" alt="Image 2020-04-27 at 11 42 46 AM" src="https://user-images.githubusercontent.com/1429108/80391866-52852080-887c-11ea-93eb-9feee512817e.png">

**Comment by garymurray** [comment]
I think let's leave this for a future cycle then as I think it has more complexities we need to try address and I don't want to rush it.

**Comment by nerrad** [comment]
Ya I think the complexity this adds is mostly around design for distinguishing global settings from block specific settings (and likely will take the most time to arrive at something that we can use consistently in other places). Technically, once we arrive at that, the implementation shouldn't be too difficult.

**Comment by haszari** [comment]
Summarising the above conversation .. 

#### Status quo / current behaviour
- There is a global store setting allowing merchants to disable coupons (`WooCommerce > Settings > General > Enable the use of coupon codes`).
- We support this in the blocks. When disabled, there's no `Add coupon` UI in the cart or checkout block.

#### Proposed enhancements (alternative options)
1. Add editor UI to the blocks allowing editing of the global `Enable coupons` setting. Suggestion to style this differently because it's not a block setting (i.e. behaves differently). 
2. Add a block setting to cart and checkout blocks allowing merchant to hide `Add coupon` UI on a per-block basis.

It looks like there was agreement to move forward with (1), but that it's possibly blocked on design. This feels risky to me, as it's an unusual pattern; needing a new design is a warning sign. Side note: I can imagine that Gutenberg & full-site editing (FSE) will hit similar issues where an **instance** of a block is the best "WYSIWYG" way to edit a **global** site option.

On the face of it, I can't see a problem with (2), and to me, this feels like the right solution for blocks. I see it working like this (apologies if this has already been discussed):

- Global store setting is still available and works the same. It's likely that merchants will always want a way to enable/disable coupons across the site.
- Each block (cart/checkout) has a block attribute for displaying the coupon UI (i.e. a regular block option, no special styling). This defaults to true; if site enables coupons globally, then the UI will appear on cart/checkout pages (automatically, no need to edit blocks).
- If the store has coupons disabled when editing a block, the coupon option is hidden. This avoids the confusion of enabling it at block level when it's globally disabled.

The use cases I can imagine for this are that a merchant wants to control where coupons can be entered, to avoid confusion/distraction during checkout. E.g. show at cart but not checkout, or vice versa. 

I can also imagine more complex scenarios where the merchant has a special checkout page which allows coupons, shared with special customers, or specific types of product etc. This is typical block approach - each block can have custom settings, can use blocks in different pages in different ways.

### Next steps
I think these are our options: 

1. Do nothing, revisit this in a future sprint or when we have more requests for flexibility with coupons. This seems reasonable, given our current goal of approximate feature parity between blocks & shortcodes. (I don't think there's a shortcode option for hiding coupon UI.)
2. Add a block option to cart & checkout to enable the coupon UI, default true.
3. Design the UI for editing the global setting (possibly in consultation with GB/FSE). Then implement it if we come up with a good solution!

@garymurray @LevinMedia @nerrad - keen for your thoughts on best next step, also cc @Aljullu

**Comment by garymurray** [comment]
> The use cases I can imagine for this are that a merchant wants to control where coupons can be entered, to avoid confusion/distraction during checkout. E.g. show at cart but not checkout, or vice versa. That seems like the driver here, and I think it's only supported by a block setting (or a pair of store settings, e.g. allow coupons in cart allow coupons in checkout).

This was the intention behind the design - to allow a merchant to decide if they wanted to enable coupons side wide, but selectively only show the field on the cart or checkout page, or both.

For now, I still think we can defer this to a future sprint/cycle.

**Comment by haszari** [comment]
FYI @nerrad @Aljullu I've moved this to backlog.

**Comment by nerrad** [comment]
Thanks Rua! I took care of removing the related metadata (cycle and epic).

**Comment by Aljullu** [comment]
@ralucaStan is this something still in the roadmap? Should we re-open the issue?

**Comment by Aljullu** [comment]
Currently, it's possible to remove the _Coupon Form_ inner block, which has the same effect. @ralucaStan I guess this issue can be closed now?

**Comment by ralucaStan** [comment]
I'd keep us open until we have a solution for settings. If one removes the coupon there's no way to add it back, unless you reinsert the block.
